### DISCUSSION TITLE
Wait time in first bottom layer ignored?

### DISCUSSION THREAD

**Discussion by msspad** [sn4k3/UVtools#402] [Q&A] [answered]
After reading [blog.honzamrazek.cz](https://blog.honzamrazek.cz/2022/01/prints-not-sticking-to-the-build-plate-layer-separation-rough-surface-on-a-resin-printer-resin-viscosity-the-common-denominator/) I wanted to test some long waiting times. I changed an imported sl1 file to "Bottom wait before cure"=40s and "Wait before cure"=5s, saved it in ctb4 and printed it on a Phrozen Mini 4k. The waiting time works fine for all layers except for the very first layer. Is this a limitation of the Chitubox firmware?

I do have 2 bottom layers: the first bottom layer gets immediately exposed once the table is in position. After retraction from the first bottom layer, the table moves into position and waits for 40s, as expected.

**Comment by sn4k3** [comment, marked as answer]
You can inspect with UVtools, it tells you the layer values per layer on "Layer Data" table.
Does first layer have the correct value?

Then it should be and sounds like a firmware restriction... They may have think: Well nothing is printed yet, so why wait long? Lets limit to a low value... We can never know what is inside the firmware without anyone inside tell us.

I can also tell you that a "**Bottom** wait before cure" does not exists on CTBv4, UVtools allow that but there is not a global property for bottom on the file, it will set to corresponding and individual layers on it own table. So first layer may is reading the global "Wait time", however the lightoff delay exists in bottom property on file format, i dont know why they think it was a cool move to not include a global Bottom for wait times...

What you can try:

1. Set both bottom and normal wait time to 40s
2. Go edit a layer range and select all normal layers.
3. Set wait time to 5s on that layer range
4. Do a dry print
5. Does it work any better?

**Reply by msspad** [reply]
Thanks for the explanation. I followed your procedure but it behaves the same: very first layer has no waiting time before cure. All other layers wait for it.
So I guess, Chitubox would have to solve this issue in their firmware...

**Reply by sn4k3** [reply]
You can try light off delay as alternative.
Use the UVtools calculator to set the light off time instead

> So I guess, Chitubox would have to solve this issue in their firmware...

They are not very responsive in this kind of stuff, it doesn't matter for them

**Reply by sn4k3** [reply]
If everything fails, you can create a empty layer at first layer with same position as second layer. That way it will count as a layer and overpass that problem. It will require scripting, UVtools will auto arrange positions if you clone layers in UI

**Reply by msspad** [reply]
- "light off delay" shows same behaviour: first layer has waiting time=0s
- then I cloned the first layer with the option "Keep the same z position for the cloned layers" enabled, cleared this cloned layer0 in the Pixel Editor and it works! 
-- "Layer Data" shows for both layer0 and layer1 PositionZ=0.05mm as expected
-- When printing, the layer0 is empty and layer1 waits correctly before exposure start.

Thanks for your help, I can now run the tests for reduced elephant foot.

**Reply by sn4k3** [reply]
> then I cloned the first layer with the option "Keep the same z position for the cloned layers" enabled, cleared this cloned layer0 in the Pixel Editor and it works!

I forget that i have added that "Keep the same z position for the cloned layers" to allow that :P

Heres another trick:
You can create a profile to blank out the layer
1. Pixel arithmetic
2. Select first layer
3. Set = All
4. Brightness = 0
5. Apply

Then with pixel editor create a pixel with brightness of 1, to prevent uvtools from clean and detect empty layer

You can also create a script to facilitate all that, but some time latter i will include this on UVtools

**Reply by msspad** [reply]
> I forget that i have added that "Keep the same z position for the cloned layers" to allow that :P

I also didn't look in the UI but found it in the code :-)


> Then with pixel editor create a pixel with brightness of 1, to prevent uvtools from clean and detect empty layer

will use "Pixel arithmetic". 
With "clean", you mean in "issues": EmptyLayer for layer0 and island warnings for layer1?

**Reply by sn4k3** [reply]
> With "clean", you mean in "issues": EmptyLayer for layer0 and island warnings for layer1?

yes for EmptyLayer, but island would still pop, UVtools not prepared to merge layers on same height for islands, a thing to improve in the future.
Also, you need to edit exposure time of that layer to a very low value to not lose time, 0.1s is ok.

With a script you can automate all this

**Reply by sn4k3** [reply]
Still in early dev, but this will work to automate all this: https://github.com/sn4k3/UVtools/blob/master/UVtools.ScriptSample/ScriptDebandingZSample.cs

You can run that via Tool -> Scripting

![image](https://user-images.githubusercontent.com/113281/149211764-286ae330-3a0d-4a9b-950e-d866fd6503b0.png)

**Reply by msspad** [reply]
Fantastic, it works perfectly. Some minor suggestions:
- I haven't heard "deband" before, maybe "build plate settlement"/settle or "build plate stabilization" works better?
- Add to "Script description": Only "safe height" is considered, "Layer range selector" is ignored. (or disable/remove "Layer range selector" what is probably not possible)

**Comment by msspad** [comment]
just noted that the bottom wait time is correctly saved but the "wait time before cure" for normal layers is always 0s in saved CTBv4 files.

**Reply by sn4k3** [reply]
It's working for me, look:

![UVtools_2022-01-13_15-37-26](https://user-images.githubusercontent.com/113281/149360730-4dfeb7cb-3c58-4913-bbe1-1a8aa086631b.png)

**Comment by msspad** [comment]
it looks fine when done in the current file. Once you save the file, close it and then reopen the ctb file, only bottom layers have the correct waiting time

**Reply by sn4k3** [reply]
Ah yes, i confirmed. I may know the problem...
Try disable light off automation: 
![image](https://user-images.githubusercontent.com/113281/149361721-4db9ddff-9eea-4249-8935-e61bf28eddb9.png)

But stills a bug on UVtools... Will fix it on next patch.

PS: Do not create new replies here, comment the current conversation instead

**Reply by msspad** [reply]
all right, thanks. 
Sorry about the reply in the wrong place, my mistake

**Reply by sn4k3** [reply]
Humm that automation is not the whole problem. It is not saving bottom wait time due the lack of that property on the file. I will fix latter.
However if you dont reload the file on UVtools you can print it and settings will be there

**Reply by msspad** [reply]
ah ok, I will try

**Reply by sn4k3** [reply]
Ok nothing wrong with UVtools actually... The problem is that i remove wait times from first blank layer. And that layer define the times for the bottom "virtual" property. That's what happen when they dont include a dam bottom property on file :( 
All you need is remove "SetNoDelays" for the first layer on the script, and it will be fine
https://github.com/sn4k3/UVtools/blob/ac16947aefbaa2401077b64eec646afb084ee900/UVtools.ScriptSample/ScriptDebandingZSample.cs#L154

Reason:
https://github.com/sn4k3/UVtools/blob/ac16947aefbaa2401077b64eec646afb084ee900/UVtools.Core/FileFormats/ChituboxFile.cs#L1360

**Reply by msspad** [reply]
Thanks for your investigation, it works without SetNoDelays()

Similar to dynamic lifts, I'm currently experimenting with dynamic "wait times before cure", see below. First layers with PositionZ<1mm always wait 40s, 
```
diff --git a/UVtools.ScriptSample/ScriptDebandingZSample.cs b/UVtools.ScriptSample/ScriptDebandingZSample.cs
index debd6cd..77d0cfa 100644
--- a/UVtools.ScriptSample/ScriptDebandingZSample.cs
+++ b/UVtools.ScriptSample/ScriptDebandingZSample.cs
@@ -137,10 +137,18 @@ namespace UVtools.ScriptSample
             if (SlicerFile.CanUseBottomWaitTimeAfterCure) SlicerFile.BottomWaitTimeAfterCure = (float) BottomWaitTimeAfterCureInput.Value;
             if (SlicerFile.CanUseWaitTimeAfterCure) SlicerFile.WaitTimeAfterCure = (float)NormalWaitTimeAfterCureInput.Value;

+            const float numPixelForMaxWaitTime = 1000000;
             foreach (var layer in SlicerFile)
             {
-                if((decimal)layer.PositionZ > BottomSafeDebandingHeightInput.Value) break;
-                layer.SetWaitTimeBeforeCureOrLightOffDelay((float) BottomWaitTimeBeforeCureInput.Value);
+                if((decimal)layer.PositionZ < BottomSafeDebandingHeightInput.Value) {
+                    layer.SetWaitTimeBeforeCureOrLightOffDelay((float) BottomWaitTimeBeforeCureInput.Value);
+                } else {
+                    var waitTimeBeforeCure = (float) NormalWaitTimeBeforeCureInput.Value;
+                    if (layer.NonZeroPixelCount < numPixelForMaxWaitTime) {
+                        waitTimeBeforeCure = waitTimeBeforeCure * layer.NonZeroPixelCount / numPixelForMaxWaitTime;
+                    }
+                    layer.SetWaitTimeBeforeCureOrLightOffDelay((float) waitTimeBeforeCure);
+                }
             }

             if (CreateEmptyLayerInput.Value)
```

Minor naming suggestion: rename button "Scripting" to "run" or "run script"

**Reply by sn4k3** [reply]
What shows here for that bottom layers? All bottoms have 40s? Or also the normal layers get that?

![UVtools_2022-01-14_17-02-07](https://user-images.githubusercontent.com/113281/149555247-2d4ac615-af58-4f0c-8410-37e6c4227673.png)

Still i think the bug is on code:

```C#
if((decimal)layer.PositionZ < BottomSafeDebandingHeightInput.Value) {
+                    layer.SetWaitTimeBeforeCureOrLightOffDelay((float) BottomWaitTimeBeforeCureInput.Value);
+                }
```
`layer.PositionZ < 1mm` is always 40s... (`BottomWaitTimeBeforeCureInput.Value`)
Your else for dynamic waits are on after 1mm
Also you are not covering the == case, it should include the 1mm into BottomWaitTimeBeforeCureInput

**Reply by msspad** [reply]
Let me explain the adapted script a bit (also for other readers):

**The issue**
PositionZ of the build plate is an open-loop control system, so the effective PositionZ also depends on external forces: 
(1) "Ground effect" force: For the first layers, the resin of the entire area of the build plate needs to be displaced.
(2) "crowded layer" force: Many pixels on the previous layer
The force of 1) is likely much higher than 2). The forces result mainly in a deformation of the printer case.

**Solution**
add WaitTimeBeforeCure to allow the build plate to settle on its PositionZ, namely ease the deformation of the printer case.

**Solution for (1)**
- (1) only occurs on the first layers. Jan's measurements indicate higher forces for about the first 32 layers. So BottomSafeDebandingHeight should be 32 * 0.05mm=1.6mm with his setup (not sure about the layer height Jan used). 
- Whether there are 31 layer or 32 layers is not critical as it's anyway just a guess that heavily depends on the setup. Using `layer.PositionZ <=` might be more consistent with UVTools and that's fine.
- All those 31 layers have WaitTimeBeforeCure=40s, no matter if they are bottom layers or normal layers.
- Bottom layers are for better adhesion to the build plate and I have typically 2 layers. So number of bottom layers is independent of number of layers with BottomWaitTimeBeforeCure=40s. We could also use another name like GroundEffectWaitTimeBeforeCure but bottom is fine for me.

**Solution for (2)**
- the force of (1) depends on the number of pixels of the previous layer. I simplify this with just using the number of pixels in the current layer and cap it at 1e6 pixels. So any layer with more than 1e6 pixels gets the full NormalWaitTimeBeforeCure and all other layer the scale between 0s and NormalWaitTimeBeforeCure. 
- The reason for the cap is that I expect the build plate to settle anyway after NormalWaitTimeBeforeCure, whether there are 1e6 or 3e6 pixels in the layer (that has to be experimentally verified as well).

**Reply by sn4k3** [reply]
Your code still have the flaw of the comparator. Should be `<=` for the first if.
Also comparing pixels are not the optimal, 1e6 pixels * 0.050mm = 50000mm², but 1e6 pixels * 0.035mm = 35000mm². Depending on printer pixel size you are set a unfair equality. Consider using a known volume and scale from there.

```C#
var layerVolume = layer.NonZeroPixelCount * Math.Max(0.05, SlicerFile.PixelSizeMax);
```

**Reply by msspad** [reply]
- if you prefer `if((decimal)layer.PositionZ <= BottomSafeDebandingHeightInput.Value) {`, that's fine. For my use cases, it does not matter.
- that's right, the area counts and not the number of pixels. 1e6 pixels is only fine for my setup and it's a hack for quick testing. The cap could be also defined with a percentage=NonZeroPixelCount/NumMaxPixel
